### DISCUSSION TITLE
Some bug in the SDK?

### DISCUSSION THREAD

**Discussion by HamzaHajeir** [earlephilhower/arduino-pico#2339] [Q&A]
Hi there,

I've recently finished porting the H4Plugins framework into running RP2040, and I'll be having a topic here soon announcing that.

In an application, everything compiles and runs perfectly, but after ~30-50 minutes, an exception/abort happens.

The code runs AsyncWebserver and MQTT. Publishing to MQTT server some simple data periodically alongwith 13kB of data every 3 seconds, and that all works well, thanks to the developed networking libraries.

An error happens after a while in a call to `rp2040.getFreeHeap()`. Have tested with two different debug flags, and I insert some information that might assist in discovering the issue.

The source code runs this with complete PlatformIO setup is availalbe in [H4Plugins Environment for Pico](https://github.com/HamzaHajeir/H4Plugins_Pico).

Any glance?

##### **CALL STACK:**
```
isr_hardfault@0x100030c4 (Unknown Source:98)
<signal handler called>@0xfffffff9 (Unknown Source:0)
__malloc_update_mallinfo@0x10096e1e (\__malloc_update_mallinfo.dbgasm:32)
_mallinfo_r@0x10096e44 (Unknown Source:3569)
mallinfo@0x100922a8 (Unknown Source:89)
RP2040::getUsedHeap@0x1001dc68 (c:\Users\Hamza\.platformio\packages\framework-arduinopico\cores\rp2040\RP2040Support.h:238)
RP2040::getFreeHeap@0x1001dc4a (c:\Users\Hamza\.platformio\packages\framework-arduinopico\cores\rp2040\RP2040Support.h:234)
_HAL_freeHeap@0x1001dda0 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4Tools\src\H4Tools.cpp:106)
_HAL_maxHeapBlock@0x1001de9a (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4Tools\src\H4Tools.cpp:112)
getMaxPayloadSize@0x10042546 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4Tools\src\H4Tools.h:115)
PublishPacket::PublishPacket@0x10044682 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4AsyncMQTT\src\Packet.cpp:448)
operator()@0x1002f276 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4AsyncMQTT\src\H4AsyncMQTT.cpp:1038)
std::__invoke_impl<short unsigned int, H4AsyncMQTT::publish(char const*, const uint8_t*, size_t, uint8_t, H4AMC_PublishOptions)::<lambda()>&>(std::__invoke_other, struct {...} &)@0x1003de80 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\invoke.h:61)
std::__invoke_r<long unsigned int, H4AsyncMQTT::publish(char const*, const uint8_t*, size_t, uint8_t, H4AMC_PublishOptions)::<lambda()>&>(struct {...} &)@0x1003a35e (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\invoke.h:114)
std::_Function_handler<long unsigned int(), H4AsyncMQTT::publish(char const*, const uint8_t*, size_t, uint8_t, H4AMC_PublishOptions)::<lambda()> >::_M_invoke(const std::_Any_data &)@0x10035432 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\std_function.h:290)
std::function<unsigned long ()>::operator()() const@0x10017ab6 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\std_function.h:591)
H4AsyncMQTT::_runGuard<unsigned short>(std::function<unsigned long ()>, unsigned short)@0x10031e7a (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4AsyncMQTT\src\H4AsyncMQTT.h:416)
H4AsyncMQTT::publish@0x1002f2fa (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4AsyncMQTT\src\H4AsyncMQTT.cpp:1038)
H4AsyncMQTT::publish@0x1002f354 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4AsyncMQTT\src\H4AsyncMQTT.cpp:1044)
H4AsyncMQTT::xPublish@0x10004dcc (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4AsyncMQTT\src\H4AsyncMQTT.h:479)
H4P_AsyncMQTT::publishDevice@0x1000566a (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4Plugins\src\H4P_AsyncMQTT.h:88)
```

##### **__malloc_update_mallinfo:** 
```asm
0x10096de0: 03 21           	movs	r1, #3
0x10096de2: 12 4b           	ldr	r3, [pc, #72]	@ (0x10096e2c <__malloc_update_mallinfo+76>)
0x10096de4: f0 b5           	push	{r4, r5, r6, r7, lr}
0x10096de6: 9a 68           	ldr	r2, [r3, #8]
0x10096de8: 52 68           	ldr	r2, [r2, #4]
0x10096dea: 8a 43           	bics	r2, r1
0x10096dec: 14 00           	movs	r4, r2
0x10096dee: 02 39           	subs	r1, #2
0x10096df0: 0f 2a           	cmp	r2, #15
0x10096df2: 00 dc           	bgt.n	0x10096df6 <__malloc_update_mallinfo+22>
0x10096df4: 00 21           	movs	r1, #0
0x10096df6: fe 20           	movs	r0, #254	@ 0xfe
0x10096df8: 03 27           	movs	r7, #3
0x10096dfa: 80 00           	lsls	r0, r0, #2
0x10096dfc: 1e 18           	adds	r6, r3, r0
0x10096dfe: 58 69           	ldr	r0, [r3, #20]
0x10096e00: 08 33           	adds	r3, #8
0x10096e02: 98 42           	cmp	r0, r3
0x10096e04: 0b d1           	bne.n	0x10096e1e <__malloc_update_mallinfo+62>
0x10096e06: 86 42           	cmp	r6, r0
0x10096e08: 07 d1           	bne.n	0x10096e1a <__malloc_update_mallinfo+58>
0x10096e0a: 09 4b           	ldr	r3, [pc, #36]	@ (0x10096e30 <__malloc_update_mallinfo+80>)
0x10096e0c: 59 60           	str	r1, [r3, #4]
0x10096e0e: 19 68           	ldr	r1, [r3, #0]
0x10096e10: 1a 62           	str	r2, [r3, #32]
0x10096e12: 89 1a           	subs	r1, r1, r2
0x10096e14: d9 61           	str	r1, [r3, #28]
0x10096e16: 5c 62           	str	r4, [r3, #36]	@ 0x24
0x10096e18: f0 bd           	pop	{r4, r5, r6, r7, pc}
0x10096e1a: 03 00           	movs	r3, r0
0x10096e1c: ef e7           	b.n	0x10096dfe <__malloc_update_mallinfo+30>
>> 0x10096e1e: 45 68           	ldr	r5, [r0, #4]
0x10096e20: 01 31           	adds	r1, #1
0x10096e22: bd 43           	bics	r5, r7
0x10096e24: c0 68           	ldr	r0, [r0, #12]
0x10096e26: 52 19           	adds	r2, r2, r5
0x10096e28: eb e7           	b.n	0x10096e02 <__malloc_update_mallinfo+34>
0x10096e2a: c0 46           	nop			@ (mov r8, r8)
0x10096e2c: 24 15           	asrs	r4, r4, #20
0x10096e2e: 00 20           	movs	r0, #0
0x10096e30: 88 20           	movs	r0, #136	@ 0x88
0x10096e32: 00 20           	movs	r0, #0
```

##### **REGISTERS:**
> r0=0x00000199
> r1=0x00000003
> r3=0x2001f358
> r4=0x00005178
> r5=0x2001a170
> r6=0x2000191c
> r7=0x00000003
> r8=0x2001ae50
> r9=0x00000001
> r10=0xffffffff
> r11=0xffffffff
> r12=0x20000167
> sp=0x20041878
> lr=0xfffffff9
> pc=0x100030c6
> xPSR=0x81000003
> msp=0x20041878
> psp=0xfffffffc
> primask=0x00000000
> basepri=0x00000000
> faultmask=0x00000000
> control=0x00000000

---------

##### **CALL STACK:**
```
isr_hardfault@0x100030c4 (Unknown Source:98)
<signal handler called>@0xfffffff9 (Unknown Source:0)
__malloc_update_mallinfo@0x1003d69a (\__malloc_update_mallinfo.dbgasm:32)
_mallinfo_r@0x1003d6c0 (\_mallinfo_r.dbgasm:7)
mallinfo@0x10038b24 (\mallinfo.dbgasm:7)
RP2040::getUsedHeap@0x10009c40 (c:\Users\Hamza\.platformio\packages\framework-arduinopico\cores\rp2040\RP2040Support.h:238)
RP2040::getFreeHeap@0x10009c40 (c:\Users\Hamza\.platformio\packages\framework-arduinopico\cores\rp2040\RP2040Support.h:234)
_HAL_freeHeap@0x10009c74 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4Tools\src\H4Tools.cpp:106)
_HAL_maxHeapBlock@0x10009c94 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4Tools\src\H4Tools.cpp:112)
operator()@0x10005d08 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\src\main.cpp:289)
std::__invoke_impl<void, onMQTTConnect()::<lambda()>&>@0x10005d36 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\invoke.h:60)
std::__invoke_r<void, onMQTTConnect()::<lambda()>&>@0x10005d36 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\invoke.h:111)
std::_Function_handler<void(), onMQTTConnect()::<lambda()> >::_M_invoke(const std::_Any_data &)@0x10005d36 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\std_function.h:290)
std::function<void ()>::operator()() const@0x10007974 (c:\users\hamza\.platformio\packages\toolchain-rp2040-earlephilhower\arm-none-eabi\include\c++\12.3.0\bits\std_function.h:591)
task::operator()@0x10008996 (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4\src\H4.cpp:183)
H4::loop@0x10008a4e (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4\src\H4.cpp:426)
loop@0x10008a7c (c:\Users\Hamza\Documents\PlatformIO\Projects\RPi Pico\.pio\libdeps\pico\H4\src\H4.cpp:364)
main@0x100254f6 (c:\Users\Hamza\.platformio\packages\framework-arduinopico\cores\rp2040\main.cpp:150)
```

##### **__malloc_update_mallinfo:**
```asm
0x1003d65c: 03 21           	movs	r1, #3
0x1003d65e: 12 4b           	ldr	r3, [pc, #72]	@ (0x1003d6a8 <__malloc_update_mallinfo+76>)
0x1003d660: f0 b5           	push	{r4, r5, r6, r7, lr}
0x1003d662: 9a 68           	ldr	r2, [r3, #8]
0x1003d664: 52 68           	ldr	r2, [r2, #4]
0x1003d666: 8a 43           	bics	r2, r1
0x1003d668: 14 00           	movs	r4, r2
0x1003d66a: 02 39           	subs	r1, #2
0x1003d66c: 0f 2a           	cmp	r2, #15
0x1003d66e: 00 dc           	bgt.n	0x1003d672 <__malloc_update_mallinfo+22>
0x1003d670: 00 21           	movs	r1, #0
0x1003d672: fe 20           	movs	r0, #254	@ 0xfe
0x1003d674: 03 27           	movs	r7, #3
0x1003d676: 80 00           	lsls	r0, r0, #2
0x1003d678: 1e 18           	adds	r6, r3, r0
0x1003d67a: 58 69           	ldr	r0, [r3, #20]
0x1003d67c: 08 33           	adds	r3, #8
0x1003d67e: 98 42           	cmp	r0, r3
0x1003d680: 0b d1           	bne.n	0x1003d69a <__malloc_update_mallinfo+62>
0x1003d682: 86 42           	cmp	r6, r0
0x1003d684: 07 d1           	bne.n	0x1003d696 <__malloc_update_mallinfo+58>
0x1003d686: 09 4b           	ldr	r3, [pc, #36]	@ (0x1003d6ac <__malloc_update_mallinfo+80>)
0x1003d688: 59 60           	str	r1, [r3, #4]
0x1003d68a: 19 68           	ldr	r1, [r3, #0]
0x1003d68c: 1a 62           	str	r2, [r3, #32]
0x1003d68e: 89 1a           	subs	r1, r1, r2
0x1003d690: d9 61           	str	r1, [r3, #28]
0x1003d692: 5c 62           	str	r4, [r3, #36]	@ 0x24
0x1003d694: f0 bd           	pop	{r4, r5, r6, r7, pc}
0x1003d696: 03 00           	movs	r3, r0
0x1003d698: ef e7           	b.n	0x1003d67a <__malloc_update_mallinfo+30>
>> 0x1003d69a: 45 68           	ldr	r5, [r0, #4]
0x1003d69c: 01 31           	adds	r1, #1
0x1003d69e: bd 43           	bics	r5, r7
0x1003d6a0: c0 68           	ldr	r0, [r0, #12]
0x1003d6a2: 52 19           	adds	r2, r2, r5
0x1003d6a4: eb e7           	b.n	0x1003d67e <__malloc_update_mallinfo+34>
0x1003d6a6: c0 46           	nop			@ (mov r8, r8)
0x1003d6a8: f4 14           	asrs	r4, r6, #19
0x1003d6aa: 00 20           	movs	r0, #0
0x1003d6ac: 40 20           	movs	r0, #64	@ 0x40
0x1003d6ae: 00 20           	movs	r0, #0
```

##### **REGISTERS:**
> r0=0x69747075
> r1=0x00000003
> r2=0x200209d8
> r3=0x200014fc
> r4=0x00006ed8
> r5=0x20019a88
> r6=0x200018ec
> r7=0x00000003
> r8=0xffffffff
> r9=0xffffffff
> r10=0xffffffff
> r11=0xffffffff
> r12=0x2000014f
> sp=0x20041e80
> lr=0xfffffff9
> pc=0x100030c4
> xPSR=0x21000003
> msp=0x20041e80
> psp=0xfffffffc
> primask=0x00000000
> basepri=0x00000000
> faultmask=0x00000000
> control=0x00000000

**Comment by HamzaHajeir** [comment]
UPDATE: Initial investigation would point out to the WDT, which should being reset every second. While it might be a missed reset, it also might be a blocked call to `rp2040.getUsedHeap()`. I'll update at soonest.

**Reply by HamzaHajeir** [reply]
UPDATE: Dismiss this comment, as it happened again with WDT completely inactivated.

**Reply by earlephilhower** [reply]
FWIW, WDT timeout would just reset the chip, not affect program behavior.  So no stack traces or panics, just a restart.

The crash looks like memory corruption in the app.  Something using after a `free` or going past the end of an allocation.  That could mess up the internal pointer structures `malloc` uses to handle free blocks (i.e. `malloc` doesn't use extra memory to manage allocations, it stores its pointer chains in unallocated regions).

**Reply by HamzaHajeir** [reply]
Thanks @earlephilhower for your input.

I've went into investigating any such possibility regarding the usage in the application, but failed in finding any issue. Saying that while having similar applications run perfectly on ESP8266/32's.

The part of the application that depend heavily on direct `malloc` and `free` is the MQTT library, which holds [memory blocks](https://github.com/HamzaHajeir/H4Tools/blob/master/src/mbx.h) of QoS1/2 packets, which gets freed when PUBACK/PUBCOMP received. No usage after that point.

However, I've collected some considerable notes regarding the issue: 
- Similar application that uses less `rp2040.getFreeHeap()` calls runs longer periods (reached 150 mins), changes goes with publishing messages other that heap and max block. And with making fewer publishes -> fewer calls to `rp2040.getFreeHeap()` would result in even longer periods (Seen 3+ hours of Uptime before just interrupting that run iteration).
- Inversely, activating debug prints in H4 Stack - which traces heap and max block on every line - as [here](https://github.com/HamzaHajeir/H4AsyncTCP/blob/8100aef2a25622dae811a0ce01acd65b781d3acc/src/H4AsyncTCP.h#L87), [here](https://github.com/HamzaHajeir/H4AsyncMQTT/blob/master/src/h4amc_common.h#L18), and [here](https://github.com/HamzaHajeir/H4Tools/blob/master/src/h4tools_config.h#L22), would result in considerably shorter periods until the issue appears (3~10 minutes), with high probability of being a call from one print. In some instances, it was the [first print](https://github.com/HamzaHajeir/H4Tools/blob/master/src/mbx.cpp#L69) after a `malloc` call.
- The issue goes in an infitine loop at `mallinfo`, I can step forever there, the shown isr hardfault signal was mostly caused by the WDT when was activated.
- Another interesting finding when H4AsyncTCP debug was on, that while that was looping in the call to `mallinfo`, the LwIP task informed the library of closing the TCP connection (because of Timout ~30 seconds), that was printed to the Serial (also heap printed out), therefore it is not kind of an exception or abort.

Theories could go in two directions:
1- It's a direct issue in `mallinfo` that might be not met by users yet, by either not having some stress on memoy, or some unusual usage pattern of buffers.
2- The H4AsyncTCP just wraps over Lwip library, which is basic C library with lots of usage of malloc/free. I can't really suspect it, but when bringing ESP32/8266 in, we might need to bring in the maintained `esp-lwip` library.

Where can we find the implementation of `mallinfo()`?

**Reply by earlephilhower** [reply]
Mallinfo is part of Newlib (4.0.0).  The code is fine and has been used by millions of embedded systems at this point, and probably a few tens of thousands of installed apps here.  The symptoms here really point to memory corruption by the app.

The chip won't immediately crash when memory is stomped on, they only crash when someone tries to access the corrupted memory.  In this case, mallinfo is actually walking the internal malloc memory structures and hitting a corrupted one and boom.  Call mallinfo more frequently, the memory will be "checked" for corruption more frequently and potentially go boom.  Don't call mallinfo at all, and you'll be fine until `malloc()` or `free()` or `new` or `delete` need to touch those bytes in RAM for some reason...

**Reply by HamzaHajeir** [reply]
Hi @earlephilhower, Thanks for the input. 

It was dense days testing and challenging the application's integrity regarding dealing with buffers, were I was mistakenly thinking it's related with the aforementioned managed memory blocks (`mbx`), where it was not even close. What also deceived me that publishing QoS1 messages ends up with the error, while QoS0 does not, therefore went through intense review to its operation. What I missed with this is QoS1 requires to receive ACK :).

The conclusion is accompanied with complete tracing of memory, which shows no single instance of usage-after-life of any pointer.

The issue goes in receiving (consecutive/multiple) data packets from the remote side. And I've successfully replicated the issue using a little modified sketch of [this one](https://github.com/HamzaHajeir/H4AsyncTCP/blob/master/examples/h4asyncCOMPARISON/h4asyncCOMPARISON.ino) in which to keep the tests run forever. Alongwith running an Echo Server on a local machine.

And with H4AsyncTCP debug on to level 3 -> more `rp2040.getFreeHeap()` calls -> the issue appears after less than a minute of running.

This brings up the lwip integration with RPiPico into the table, especially when this exact library works perfectly for ESP32/8266.

How RPI SDK integrates LwIP, and how LwIP runs by the Arduino core?

**Reply by earlephilhower** [reply]
We use the SDK's integration.  Please see the documentation on `async_context_threadsafe_background`.

**Reply by HamzaHajeir** [reply]
Then I think initially I need to acquire a lock for complete porting, correct?

**Reply by earlephilhower** [reply]
Not really sure what you're doing, but there is a lock in the context that should be acquired before LWIP ops.

All LWIP ops used in the core or included libraries are *automatically* wrapped to acquire and release the context lock, buf if you're using some new LWIP call then look at `cores/rp2040/lwip-wrap.cpp` and `lib/platfowm-wrap.txt` for the magic incantations.

**Reply by HamzaHajeir** [reply]
I see, I found that I don't need to do locking.

We're now left with the same issue. After a while the `mallinfo()` (called by `rp2040.getFreeHeap()` goes in infinite loop. 

For an observation to logs, the memory corruption can't be deemed a viable point, as after it gets in infinite loop, the async_context of lwip calls the apps `_raw_recv` and `_raw_error`, in which they call `mallinfo` implicitly to get the free heap, just like: 

```
01:35:25.846 > H4AT:1: H=181560 M=90780 _raw_recv 0x20013168 tpcb=0x2000dd00 p=0x2000d0f0 err=0 data=0x2000d136 tot_len=25
01:35:25.846 > H4AT:2: H=181536 M=90768 * p=0x2000d0f0 * FREE DATA 0x2000d136 25 0x01 bpp=0
01:35:25.846 > H4AT:2: H=181424 M=90712 pbuf_free(0x2000d0f0) -> 1
>>>>>>>>>>>>
01:35:54.469 > H4AT:1: H=181392 M=90696 _raw_recv 0x20013168 tpcb=0x2000dd00 p=0 err=0 data=0 tot_len=0
01:35:54.469 > H4AT:1: H=181424 M=90712 Will Close!
01:37:54.479 > H4AT:1: H=181312 M=90656 _raw_error c=0x20013168 e=-14
```

I've marked where it goes infinitely, ~30 seconds later it reads the FreeHeap successfully for 6 times.

What else could it be?

UPDATE: 

It's also interesting that SerialUSB stopped printing when the issue happens, wherein by reading the call stack, it shows me that there are missing prints in between. However the same SerialUSB successfully prints from within the other async_context call. 

Is it something with the async_context? I might imagine it being something (as) stack size of task. But I don't have knowledge to the internals of, currently.

**Reply by HamzaHajeir** [reply]
Hi @earlephilhower, have we reached to a conclusion this issue can't be fixed?

**Reply by earlephilhower** [reply]
Yes, you have a memory corruption issue that's not in the core, so there's nothing to do here.

**Reply by HamzaHajeir** [reply]
> Yes, you have a memory corruption issue that's not in the core, so there's nothing to do here.

If that's true that would be an obvious issue in other platforms, besides that `mallinfo` would always fail after the first one, which is not.
